File: Data Structures & Algorithms/eating-bananas/submission-2.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int upper = 0;
        for(int i : piles)
        {
            upper = max(upper,i);
        }

        int r = upper;
        int l = 1;
        int m;
        int time_taken = 0;
        int res;

        while(l <= r)
        {
            m = l + (r-l)/2;
            time_taken = 0;

            for(int i : piles)
            {
                time_taken += (i + m - 1) / m;
            }

            if (time_taken > h)
            {
                l = m+1;
            }
            else
            {
                res = m;
                r = m-1;
            }
        }

        return res;
    }
};

File: Data Structures & Algorithms/car-fleet/submission-2.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        res = 1
        n = len(position)
        cars = sorted(zip(position, speed), reverse = True)
        stack = []
        stack.append(cars[0])
        for i in range(len(cars)):
            time_taken = (target - cars[i][0]) / cars[i][1]
            time_taken_b = (target - stack[0][0]) / stack[0][1]
            if time_taken > time_taken_b:
                res += 1
                stack.clear()
                stack.append(cars[i])

        return res

